### PR TITLE
etcdctlv3: add initial mirrormaker

### PR DESCRIPTION
**Pull request by xiang90**
Fix https://github.com/coreos/etcd/issues/1749

This is an initial implementation, do not expect this to be prefect.

### CODE REVIEW

**Comment by gyuho** [comment]
lgtm.

File: etcdctlv3/command/make_mirror_command.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	// TODO: remove the prefix of the destination cluster?
+	dkv := clientv3.NewKV(dc)
+
+	s := sync.NewSyncer(c, mmprefix, 0)
+
+	rc, errc := s.SyncBase(context.TODO())
+
+	for r := range rc {
+		for _, kv := range r.Kvs {
+			_, err := dkv.Put(ctx, string(kv.Key), string(kv.Value))
```
Comment (heyitsanthony) [inline etcdctlv3/command/make_mirror_command.go]: does it matter that this clobbers revisions?

File: etcdctlv3/command/make_mirror_command.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	wc := s.SyncUpdates(context.TODO())
+
+	for wr := range wc {
+		if wr.CompactRevision != 0 {
+			return v3rpc.ErrCompacted
+		}
+
+		for _, ev := range wr.Events {
```
Comment (heyitsanthony) [inline etcdctlv3/command/make_mirror_command.go]: sequences of events with equal revisions probably need to be batched as txns

**Comment by heyitsanthony** [comment]
The exact revisions probably don't need to be preserved, but maintaining the revision ordering on the keys may be important for correctness.

**Comment by xiang90** [comment]
@heyitsanthony I am less worried about preserving the initial revisions and atomicity, since the kv state is only consistency + usable after the full sync (we have to make the base in order and some ops in txn). But I do agree we can do better on the update side. Update the code accordingly, please take a look.

**Comment by heyitsanthony** [comment]
@xiang90 I think the revision ordering can still matter. For instance, dropping the revision ordering will shuffle queues. An example that might be catastrophic is an LFS-style multiple super-block type of algorithm where the most up-to-date checkpoint is chosen based on mod revision (not sure how applicable it is here).

**Comment by xiang90** [comment]
> For instance, dropping the revision ordering will shuffle queues.

For mirror, this is not huge a problem. Mirror is readonly, and I do not expect there is a actual consumer of the queue item at the destination side. There might be a observer of the queue, but it will observe the correct consuming order. 

After a second though, this actually applies to our current snapshot implementation. But do you have any scalable solution in mind? Does this mean that we have to keep intermediate data somehow on disk? Or for snapshot, we do not keep these "dynamic" data structure working after recovery?

**Comment by heyitsanthony** [comment]
The etcdctlv3 snapshot dumps revision data so it's OK unless the restore drops the revisions.

I can think of two (not so ideal) solutions. Most obvious is write out the base somewhere and put into the mirror sorted by mod revision. Second would be sorting by mod revision in  `SyncBase`'s `Range` queries, but the sort implementation right now may be too slow for that.

The creation revision might be important too, but that needs two passes to get the ordering right. Supporting version would be even worse.

I suspect this can't be done adequately with plain `Put`s.

One possible mirroring criteria could be that any txn that would have succeeded (or failed) on the primary cluster on the base revision will succeed (or fail) on the mirror on the base revision. This is stricter than only maintaining ordering since it needs exact revisions and versions, but it has better replay properties.

**Comment by xiang90** [comment]
> One possible mirroring criteria could be that any txn that 

This is hard since the destination cluster might have other inputs. The mirrormaker is just a part of that inputs. This is the same challenge for constructing the base. Probably we should just accept the fact and do not make any guarantees on the base and only make ordering guarantees on the update (not revision).

**Comment by heyitsanthony** [comment]
OK, I guess we can get away with the mirror clobbering the base metadata so long as snapshot restore keeps the metadata intact.
lgtm

**Comment by heyitsanthony** [comment]
One last thing about this: what about leases?

**Comment by xiang90** [comment]
@heyitsanthony Leases will not be mirrored. We basically only make a readonly data mirror. All the active will still happen in the source side. If the key is deleted by a lease expire, it will reflect on the other side.

**Comment by heyitsanthony** [comment]
OK

**Comment by xiang90** [comment]
@heyitsanthony I will write documentation to reflect our discussion for mirrormaker.

**Pull request merged**